images/converter/default.go

```diff
@@ -127,6 +140,23 @@ func (c *defaultConverter) convert(ctx context.Context, cs content.Store, desc o
 	return newDesc, nil
 }
 
+// singleflight makes sure that only one conversion execution is
+// in-flight for a given descriptor at a time.
+func (c *defaultConverter) singleflight(ctx context.Context, cs content.Store, desc ocispec.Descriptor, convert ConvertFunc) (*ocispec.Descriptor, error) {
+	if c.sg == nil {
+		return convert(ctx, cs, desc)
+	}
+
+	newDesc, err, _ := c.sg.Do(desc.Digest.String(), func() (interface{}, error) {
```

Comment:
Why properties other than digest aren't included to the singleflight key? Converters may be aware of them and produce different results.

Reply:
Thanks for the reply, made a key with `remotes.MakeRefKey`.

Reply:
If there is a performance tradeoff, I think we should make it optional feature.

Reply:
```
func makeDesc(i int) ocispec.Descriptor {
	return ocispec.Descriptor{
		MediaType: ocispec.MediaTypeImageLayer,
		Digest:    digest.FromBytes([]byte(fmt.Sprintf("%d", i))),
		Size:      100,
		URLs:      []string{},
		Annotations: map[string]string{
			"foo": "baz",
			"baz": "boo",
		},
		Data:     []byte{},
		Platform: &ocispec.Platform{},
	}
}

func makeRefKey(ctx context.Context, desc ocispec.Descriptor) string {
	key := desc.Digest.String()
	if desc.Annotations != nil {
		for k, v := range desc.Annotations {
			key = fmt.Sprintf("%s.%s=%s", key, k, v)
		}
	}

	return fmt.Sprintf("%s.%s", key, desc.MediaType)
}

func makeJSONKey(ctx context.Context, desc ocispec.Descriptor) []byte {
	bytes, _ := json.Marshal(desc)
	return bytes
}

func BenchmarkMakeRefKey(b *testing.B) {
	keyMap := map[string]string{}
	for n := 0; n < b.N; n++ {
		key := makeRefKey(context.TODO(), makeDesc(n))
		keyMap[key] = key
	}
	for n := 0; n < b.N; n++ {
		for k, v := range keyMap {
			require.Equal(b, k, v)
		}
	}
}

func BenchmarkMakeJSONKey(b *testing.B) {
	keyMap := map[string]string{}
	for n := 0; n < b.N; n++ {
		makeJSONKey(context.TODO(), makeDesc(n))
	}
	for n := 0; n < b.N; n++ {
		for k, v := range keyMap {
			require.Equal(b, k, v)
		}
	}
}
```

```
$ go test -bench=. -benchtime=2000x images/converter

goos: linux
goarch: amd64
pkg: github.com/containerd/containerd/images/converter
cpu: Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
BenchmarkMakeRefKey-6    	    2000	   1424219 ns/op
BenchmarkMakeJSONKey-6   	    2000	      3010 ns/op
PASS
ok  	github.com/containerd/containerd/images/converter	2.865s
```

@ktock the simple bench shows that json key is faster than ref key.

Updated commits, provided optional `WithSingleflight()` for the feature in the second commit, PTAL.